src/esp/sim/Simulator.cpp

```diff
@@ -1126,132 +1054,114 @@ esp::sensor::Sensor& Simulator::addSensorToObject(
+  if (newSensor.isVisualSensor() && config_.createRenderer) {
+    renderer_->bindRenderTarget(
+        static_cast<esp::sensor::VisualSensor&>(newSensor));
+  }
+
+  return newSensor;
 }
 
 void Simulator::setPathFinder(nav::PathFinder::ptr pathfinder) {
   pathfinder_ = std::move(pathfinder);
 }
-gfx::RenderTarget* Simulator::getRenderTarget(int agentId,
-                                              const std::string& sensorId) {
-  agent::Agent::ptr ag = getAgent(agentId);
-
-  if (ag != nullptr) {
-    sensor::Sensor& sensor = ag->getSubtreeSensorSuite().get(sensorId);
-    if (sensor.isVisualSensor()) {
-      return &(static_cast<sensor::VisualSensor&>(sensor).renderTarget());
+gfx::RenderTarget* Simulator::getRenderTarget(const std::string& sensorUuid) {
+  // TODO: This should be a global map lookup instead of a tree search
+  auto allSensors = getActiveSceneGraph().getRootNode().getSubtreeSensors();
+  auto sensor = allSensors.find(sensorUuid);
+  if (sensor != allSensors.end()) {
+    if (sensor->second.get().isVisualSensor()) {
+      return &(static_cast<sensor::VisualSensor&>(sensor->second.get())
+                   .renderTarget());
     }
   }
+
   return nullptr;
 }
 
-bool Simulator::displayObservation(const int agentId,
-                                   const std::string& sensorId) {
-  agent::Agent::ptr ag = getAgent(agentId);
+bool Simulator::displayObservation(const std::string& sensorUuid) {
+  // TODO: This should be a global map lookup instead of a tree search
+  auto allSensors = getActiveSceneGraph().getRootNode().getSubtreeSensors();
+  auto sensor = allSensors.find(sensorUuid);
 
-  if (ag != nullptr) {
-    sensor::Sensor& sensor = ag->getSubtreeSensorSuite().get(sensorId);
-    return sensor.displayObservation(*this);
+  if (sensor != allSensors.end()) {
+    return sensor->second.get().displayObservation(*this);
   }
   return false;
 }
 
-bool Simulator::drawObservation(const int agentId,
-                                const std::string& sensorId) {
-  agent::Agent::ptr ag = getAgent(agentId);
+bool Simulator::drawObservation(const std::string& sensorUuid) {
+  // TODO: This should be a global map lookup instead of a tree search
+  auto allSensors = getActiveSceneGraph().getRootNode().getSubtreeSensors();
+  auto sensor = allSensors.find(sensorUuid);
 
-  if (ag != nullptr) {
-    sensor::Sensor& sensor = ag->getSubtreeSensorSuite().get(sensorId);
-    if (sensor.isVisualSensor()) {
-      return static_cast<sensor::VisualSensor&>(sensor).drawObservation(*this);
+  if (sensor != allSensors.end()) {
+    if (sensor->second.get().isVisualSensor()) {
+      return static_cast<sensor::VisualSensor&>(sensor->second.get())
+          .drawObservation(*this);
     }
   }
   return false;
 }
 
-bool Simulator::visualizeObservation(int agentId, const std::string& sensorId) {
-  agent::Agent::ptr ag = getAgent(agentId);
+bool Simulator::visualizeObservation(const std::string& sensorUuid) {
```

Comment:
Minor thing, but there's a naming inconsistency with the observation-based functions.  If we plan on ever supporting non-sensor-based observations, then Simulator::drawObservation and Simulator::visualizeObservation should probably be renamed to drawSensorObservation and visualizeSensorObservation, respectively.

If we do not ever plan on having non-sensor-driven observations, then getSensorObservation should probably just be getObservation

Reply:
Good point for a follow-up. I'll be touching these again with the next phase of refactor. :+1: